datafusion/expr/src/logical_plan/invariants.rs

```diff
@@ -41,19 +61,54 @@ pub enum InvariantLevel {
     Executable,
 }
 
+/// Apply the [`InvariantLevel::Always`] check at the root plan node only.
 pub fn assert_always_invariants(plan: &LogicalPlan) -> Result<()> {
     // Refer to <https://datafusion.apache.org/contributor-guide/specification/invariants.html#relation-name-tuples-in-logical-fields-and-logical-columns-are-unique>
     assert_unique_field_names(plan)?;
 
     Ok(())
 }
 
+/// Visit the plan nodes, and confirm the [`InvariantLevel::Executable`]
+/// as well as the less stringent [`InvariantLevel::Always`] checks.
 pub fn assert_executable_invariants(plan: &LogicalPlan) -> Result<()> {
```

Comment:
Documenting the existing behavior.
* The `assert_always_invariants()` assess only the current node, and does not assess the remaining DAG. 
* whereas the `assert_executable_invariants()` (a) visits the subplan, and (b) validates the always and executable.

Reply:
The does feel like a blurring of definitions, but the previous decision was made based upon performance. Maybe some re-naming would help? 🤔